### ISSUE TITLE
Random 502 responses on OSX

### ISSUE THREAD

**Issue by bstst** [nghttp2/nghttp2#563]
Issue: Almost always getting 502 responses. Sometimes it goes through though. If I set log-level to INFO -- I get much less 502s, but still get them very consistently, e.g. first load is almost always a 502, a consecutive load is almost always a 200, but then when connection times out in the log (`01/Apr/2016:10:50:29 +0300 PID86810 [INFO] shrpx_http_downstream_connection.cc:527 [DCONN:0x7fa29a702070] Idle connection timeout`) -- the next request is almost always a 502.

Most often the error in the log is this:
`01/Apr/2016:10:58:29 +0300 PID87318 [INFO] shrpx_http2_upstream.cc:1099 [DCONN:0x7fde6250ca90] Downstream network/general error`

And sometimes it's this one:
`01/Apr/2016:10:32:24 +0300 PID85467 [INFO] shrpx_https_upstream.cc:780 [DCONN:0x7fb778c1dad0] Network error/general error`

Am I doing something wrong?

Environment:
OSX El Capitan 10.11.3, nghttpx nghttp2/1.9.1

Command:
`nghttpx --conf nghttpx.conf`

Config:
```
frontend=127.0.0.1,443
backend=127.0.0.1,80
cacert=server.csr
private-key-file=server.key
certificate-file=server.crt
add-x-forwarded-for=yes
no-via=yes
no-ocsp=yes
insecure=yes
```

**Comment by tatsuhiro-t** [comment]
There is several cases when 502 is returned from nghttpx.
If backend server returns content-length header which includes the value that cannot be parsed as integer, nghttpx will return 502.

If you can, could you share the relevant portion of nghttpx INFO log?  Also we are interested in the response header fields the backend server returned.

**Comment by bsuh** [comment]
I am also getting this on FreeBSD. So it might be BSD specific.

**Comment by tatsuhiro-t** [comment]
Thank you.  Since I don't have this issue on Linux, it is logical to think that this is *BSD specific issue.
BTW, the log cited in the previous comment is pretty normal, and does not describe 502 situation.

**Comment by tatsuhiro-t** [comment]
I have old FreeBSD (2 years ago), and compiled nghttp2 master branch.  I encountered gcc49 bug, but after nghttpx was built, I have no issue regarding 502 response.

**Comment by bsuh** [comment]
I was using nghttp2 1.9.1 (latest from FreeBSD package repository). And it seems like this issue was opened before 1.9.2 release. Maybe it's not a bug in 1.9.2/master.

**Comment by bsuh** [comment]
Still occurred on 1.9.2 installed from FreeBSD ports.
Server
```
vagrant@:/usr/ports/www/nghttp2 % sudo nghttpx -L INFO -f0.0.0.0,443 -b127.0.0.1,9999 /usr/local/etc/nginx/bsuh.io.key /usr/local/etc/nginx/bundle.crt
23/Apr/2016:14:00:33 +0000 PID30928 [INFO] shrpx.cc:2134 Reverse sorted wildcard hosts (compared from tail to head, and sorted in reverse order):
23/Apr/2016:14:00:33 +0000 PID30928 [INFO] shrpx.cc:2143 Resolving backend address
23/Apr/2016:14:00:33 +0000 PID30928 [INFO] shrpx.cc:2153 Host-path pattern: group 0: '/', proto=http/1.1
23/Apr/2016:14:00:33 +0000 PID30928 [INFO] shrpx.cc:2156 group 0 -> 127.0.0.1:9999
23/Apr/2016:14:00:33 +0000 PID30928 [INFO] shrpx.cc:2170 Catch-all pattern is group 0
23/Apr/2016:14:00:33 +0000 PID30928 [INFO] shrpx.cc:183 Address resolution for 127.0.0.1 succeeded: 127.0.0.1
23/Apr/2016:14:00:33 +0000 PID30928 [NOTICE] Resolved backend address: 127.0.0.1:9999 -> 127.0.0.1:9999
23/Apr/2016:14:00:33 +0000 PID30928 [NOTICE] Listening on 0.0.0.0, tls
23/Apr/2016:14:00:33 +0000 PID30928 [NOTICE] Worker process [30929] spawned
23/Apr/2016:14:00:33 +0000 PID30928 [NOTICE] Renew new ticket keys
23/Apr/2016:14:00:33 +0000 PID30928 [INFO] shrpx_worker_process.cc:182 enc_keylen=16, hmac_keylen=32
23/Apr/2016:14:00:33 +0000 PID30928 [INFO] shrpx_worker_process.cc:241 ticket keys generation done
23/Apr/2016:14:00:33 +0000 PID30928 [INFO] shrpx_worker_process.cc:243 0 enc+dec: 7853db0449e2f59482e6e2235162a494
23/Apr/2016:14:00:33 +0000 PID30928 [INFO] shrpx_connection_handler.cc:457 Start ocsp update for /usr/local/etc/nginx/bundle.crt
23/Apr/2016:14:00:33 +0000 PID30928 [INFO] shrpx_worker_process.cc:528 Entering event loop
env: python: No such file or directory
23/Apr/2016:14:00:33 +0000 PID30928 [WARN] shrpx_connection_handler.cc:605 ocsp query command for /usr/local/etc/nginx/bundle.crt failed: error=0, rstatus=32512, status=127
23/Apr/2016:14:00:45 +0000 PID30928 [INFO] shrpx_connection_handler.cc:332 [LISTEN:0x7fffffffb400] Accepted connection. fd=7
23/Apr/2016:14:00:45 +0000 PID30928 [INFO] shrpx_connection.cc:418 tls: handshake is still in progress
23/Apr/2016:14:00:45 +0000 PID30928 [INFO] shrpx_connection.cc:418 tls: handshake is still in progress
23/Apr/2016:14:00:45 +0000 PID30928 [INFO] shrpx_connection.cc:418 tls: handshake is still in progress
23/Apr/2016:14:00:45 +0000 PID30928 [INFO] shrpx_connection.cc:418 tls: handshake is still in progress
23/Apr/2016:14:00:45 +0000 PID30928 [INFO] shrpx_connection.cc:469 SSL/TLS handshake completed
23/Apr/2016:14:00:45 +0000 PID30928 [INFO] shrpx_client_handler.cc:203 [CLIENT_HANDLER:0x8034b0000] SSL/TLS handshake completed
23/Apr/2016:14:00:45 +0000 PID30928 [INFO] shrpx_client_handler.cc:535 [CLIENT_HANDLER:0x8034b0000] No protocol negotiated. Fallback to HTTP/1.1
23/Apr/2016:14:00:45 +0000 PID30928 [INFO] shrpx_https_upstream.cc:69 [UPSTREAM:0x803417740] HTTP request started
23/Apr/2016:14:00:45 +0000 PID30928 [INFO] shrpx_https_upstream.cc:285 [UPSTREAM:0x803417740] HTTP request headers completed
23/Apr/2016:14:00:45 +0000 PID30928 [INFO] shrpx_https_upstream.cc:312 [UPSTREAM:0x803417740] HTTP request headers
GET / HTTP/1.1
host: bsuh.io
user-agent: curl/7.43.0
accept: */*

23/Apr/2016:14:00:45 +0000 PID30928 [INFO] shrpx_client_handler.cc:722 [CLIENT_HANDLER:0x8034b0000] Downstream address group_idx: 0
23/Apr/2016:14:00:45 +0000 PID30928 [INFO] shrpx_client_handler.cc:733 [CLIENT_HANDLER:0x8034b0000] Downstream connection pool is empty. Create new one
23/Apr/2016:14:00:45 +0000 PID30928 [INFO] shrpx_http_downstream_connection.cc:135 [DCONN:0x80347b980] Attaching to DOWNSTREAM:0x803486900
23/Apr/2016:14:00:45 +0000 PID30928 [INFO] shrpx_http_downstream_connection.cc:224 [DCONN:0x80347b980] Connecting to downstream server
23/Apr/2016:14:00:45 +0000 PID30928 [INFO] shrpx_http_downstream_connection.cc:455 [DCONN:0x80347b980] HTTP request headers. stream_id=0
GET / HTTP/1.1
Host: bsuh.io
User-Agent: curl/7.43.0
Accept: */*
X-Forwarded-Proto: https
Via: 1.1 nghttpx


23/Apr/2016:14:00:45 +0000 PID30928 [INFO] shrpx_https_upstream.cc:439 [UPSTREAM:0x803417740] HTTP request completed
23/Apr/2016:14:00:45 +0000 PID30928 [INFO] shrpx_http_downstream_connection.cc:1065 [DCONN:0x80347b980] Connected to downstream host
23/Apr/2016:14:00:45 +0000 PID30928 [INFO] shrpx_https_upstream.cc:780 [DCONN:0x80347b980] Network error/general error
23/Apr/2016:14:00:45 +0000 PID30928 [INFO] shrpx_downstream.cc:160 [DOWNSTREAM:0x803486900] Deleting
23/Apr/2016:14:00:45 +0000 PID30928 [INFO] shrpx_downstream.cc:190 [DOWNSTREAM:0x803486900] Deleted
23/Apr/2016:14:00:45 +0000 PID30928 [INFO] shrpx_client_handler.cc:468 [CLIENT_HANDLER:0x8034b0000] Deleting
23/Apr/2016:14:00:45 +0000 PID30928 [INFO] shrpx_client_handler.cc:491 [CLIENT_HANDLER:0x8034b0000] Deleted
```

Curl
```
* Rebuilt URL to: https://bsuh.io/
*   Trying 192.168.33.10...
* Connected to bsuh.io (192.168.33.10) port 443 (#0)
* TLS 1.2 connection using TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256
* Server certificate: bsuh.io
* Server certificate: COMODO RSA Domain Validation Secure Server CA
* Server certificate: COMODO RSA Certification Authority
* Server certificate: AddTrust External CA Root
> GET / HTTP/1.1
> Host: bsuh.io
> User-Agent: curl/7.43.0
> Accept: */*
>
< HTTP/1.1 502 Bad Gateway
< Server: nghttpx nghttp2/1.9.2
< Content-Length: 143
< Date: Sat, 23 Apr 2016 14:00:45 GMT
< Content-Type: text/html; charset=UTF-8
< Connection: close
<
* Closing connection 0
<!DOCTYPE html><html lang="en"><title>502 Bad Gateway</title><body><h1>502 Bad Gateway</h1><footer>nghttpx nghttp2/1.9.2</footer></body></html>
```

I will try master branch next.

**Comment by tatsuhiro-t** [comment]
Thank you.  It looks like we get error while writing to or reading from backend server.
I made a following patch to log the syscall error when reading/writing backend.
Could you apply this patch, and get the log again?

```diff
diff --git a/src/shrpx_connection.cc b/src/shrpx_connection.cc
index 8a5814b..e4ebab6 100644
--- a/src/shrpx_connection.cc
+++ b/src/shrpx_connection.cc
@@ -667,6 +667,8 @@ ssize_t Connection::write_clear(const void *data, size_t len) {
       ev_timer_again(loop, &wt);
       return 0;
     }
+    auto error = errno;
+    LOG(WARN) << "DEBUG: write(2) returned error " << strerror(error);
     return SHRPX_ERR_NETWORK;
   }
 
@@ -690,6 +692,8 @@ ssize_t Connection::writev_clear(struct iovec *iov, int iovcnt) {
       ev_timer_again(loop, &wt);
       return 0;
     }
+    auto error = errno;
+    LOG(WARN) << "DEBUG: writev(2) returned error " << strerror(error);
     return SHRPX_ERR_NETWORK;
   }
 
@@ -711,6 +715,8 @@ ssize_t Connection::read_clear(void *data, size_t len) {
     if (errno == EAGAIN || errno == EWOULDBLOCK) {
       return 0;
     }
+    auto error = errno;
+    LOG(WARN) << "DEBUG: read(2) returned error " << strerror(error);
     return SHRPX_ERR_NETWORK;
   }
 
```

**Comment by bsuh** [comment]
```
23/Apr/2016:14:48:54 +0000 PID64363 [INFO] shrpx_https_upstream.cc:439 [UPSTREAM:0x803417740] HTTP request completed
23/Apr/2016:14:48:54 +0000 PID64363 [INFO] shrpx_http_downstream_connection.cc:1065 [DCONN:0x80347b980] Connected to downstream host
23/Apr/2016:14:48:54 +0000 PID64363 [WARN] shrpx_connection.cc:696 DEBUG: writev(2) returned error Socket is not connected
23/Apr/2016:14:48:54 +0000 PID64363 [INFO] shrpx_https_upstream.cc:780 [DCONN:0x80347b980] Network error/general error
23/Apr/2016:14:48:54 +0000 PID64363 [INFO] shrpx_downstream.cc:160 [DOWNSTREAM:0x803486900] Deleting
23/Apr/2016:14:48:54 +0000 PID64363 [INFO] shrpx_downstream.cc:190 [DOWNSTREAM:0x803486900] Deleted
23/Apr/2016:14:48:54 +0000 PID64363 [INFO] shrpx_client_handler.cc:468 [CLIENT_HANDLER:0x8034b0000] Deleting
23/Apr/2016:14:48:54 +0000 PID64363 [INFO] shrpx_client_handler.cc:491 [CLIENT_HANDLER:0x8034b0000] Deleted
```

**Comment by tatsuhiro-t** [comment]
Thank you.

```
23/Apr/2016:14:48:54 +0000 PID64363 [INFO] shrpx_http_downstream_connection.cc:1065 [DCONN:0x80347b980] Connected to downstream host
23/Apr/2016:14:48:54 +0000 PID64363 [WARN] shrpx_connection.cc:696 DEBUG: writev(2) returned error Socket is not connected
```

After we checked connection was made (this was done by polling writable state, and check getsockopt passing SOL_SOCKET and SO_ERROR), and when writing something, we found that socket was not connected.

Did you see the backend got SYN packet from nghttpx?  We'd like to make sure that the connection was made between nghttpx and backend.  If backend suddenly dropped connection, then probably it sees something it does not like.  Probably, we need to capture packets between them for this.

**Comment by bsuh** [comment]
If you look at the logs, the HTTP request completed to upstream so definitely a connection was made. I am not sure why it's trying to write to upstream any more after that.

**Comment by tatsuhiro-t** [comment]
In nghttpx term, upstream means client, and downstream is backend.  Sorry for confusion.
We create HTTP request to backend before connecting to the backend server.

> 23/Apr/2016:14:00:45 +0000 PID30928 [INFO] shrpx_http_downstream_connection.cc:224 [DCONN:0x80347b980] Connecting to downstream server

So the following log means that nghttpx received whole HTTP request from client:

> 23/Apr/2016:14:48:54 +0000 PID64363 [INFO] shrpx_https_upstream.cc:439 [UPSTREAM:0x803417740] HTTP request completed

And nghttpx thought that connection to backend was made:

> 23/Apr/2016:14:48:54 +0000 PID64363 [INFO] shrpx_http_downstream_connection.cc:1065 [DCONN:0x80347b980] Connected to downstream host

But it turned out that socket was closed:

> 23/Apr/2016:14:48:54 +0000 PID64363 [WARN] shrpx_connection.cc:696 DEBUG: writev(2) returned error Socket is not connected




  Perhaps this is the log you read.  In the log "

**Comment by bsuh** [comment]
It looks like nghttp2 sends RST or FIN to backend for some reason.

![image](https://cloud.githubusercontent.com/assets/47231/14762433/f3e3bd00-092e-11e6-8682-0a2f9d03572c.png)

![image](https://cloud.githubusercontent.com/assets/47231/14762435/012f2652-092f-11e6-81da-1abaf128aeff.png)

[pcap.zip](https://github.com/nghttp2/nghttp2/files/233017/pcap.zip)

**Comment by tatsuhiro-t** [comment]
Thank you for pcap.
It is very interesting that nghttpx sends FIN without sending any HTTP request.
At the moment, I'm not sure why this happens.

As with other event based client implementation, nghttpx uses nonblocking connect to remote server.
nghttpx first makes a nonblocking socket, and issues connect(2), which most likely to return EINPROGRESS.  Then we wait until socket is writable.  We use libev for this (and I was a bit surprised when I knew that kqueue is not used by default unless OS is not NetBSD).
After OS notified us that socket is writable, we use getsockopt(2) to get pending socket error.  If there is no error, then we assume that connection was established properly, and start to write HTTP request to backend.  In FreeBSD case, getsockopt(2) indicates no error, and issues writev(2), but it failed with errno ENOTCONN.

Here is strace result (only relevant portion of the log is shown):

```
[pid  4873] socket(PF_INET, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, IPPROTO_IP) = 14
[pid  4873] setsockopt(14, SOL_TCP, TCP_NODELAY, [1], 4) = 0
[pid  4873] connect(14, {sa_family=AF_INET, sin_port=htons(80), sin_addr=inet_addr("127.0.0.1")}, 16) = -1 EINPROGRESS (Operation now in progress)
...
[pid  4873] getsockopt(14, SOL_SOCKET, SO_ERROR, [0], [4]) = 0
[pid  4873] writev(14, [{"GET / HTTP/1.1\r\nHost: localhost:"..., 159}], 1) = 159
```

I found that we ignored the return value of getsockopt() in nghttpx code, and fixed it in the master branch.

If notification of writable socket is spurious, then we might do this before connection is established.  But I'm not sure why that happens on FreeBSD.

**Comment by tatsuhiro-t** [comment]
FYI, on fresh install FreeBSD 10.3R with all dependencies installed via pkg, I could not reproduce this issue.

**Comment by bsuh** [comment]
It seems like a race condition. When I use FreeBSD equivalent of strace, truss, and it doesn't seem to happen. Also it happens less often with log level set to INFO. So maybe nghttpx is doing some operation too quickly unless blocked by some IO.

**Comment by bsuh** [comment]
Able to trigger error using dtruss
```
 19246/100087:  socket(0x2, 0x30000001, 0x0)		 = 12 0
 19246/100087:  setsockopt(0xC, 0x6, 0x1)		 = 0 0
 19246/100087:  connect(0xC, 0x803495B48, 0x10)		 = -1 Err#36
 19246/100087:  read(0x7, "\0", 0x5)		 = -1 Err#35
 19246/100087:  getsockopt(0xC, 0xFFFF, 0x1007)		 = 0 0
 19246/100087:  writev(0xC, 0x7FFFFFFFAAF8, 0x1)		 = -1 Err#57
 19246/100087:  gettimeofday(0x7FFFFFFFA930, 0x0, 0x0)		 = 0 0
 19246/100087:  shutdown(0xC, 0x1, 0x0)		 = 0 0
 19246/100087:  close(0xC)		 = 0 0
 19246/100087:  clock_gettime(0x4, 0x7FFFFFFFAC70, 0x0)		 = 0 0
 19246/100087:  kevent(0x9, 0x803420A80, 0x3)		 = 1 0
 19246/100087:  clock_gettime(0x4, 0x7FFFFFFFAC70, 0x0)		 = 0 0
 19246/100087:  getpid(0x0, 0x0, 0x0)		 = 19246 0
 19246/100087:  write(0x7, "\027\003\003\001\200m=j\221Lz\362Y/}\241\275F\302\233\006N\\\204\027\372\211\272\246\020N\310\320\363_v\020L\003\302I\214\325k\026h\364_\275\322\207\337\212\321\222\310\203\3524\233+\323\375#\037\377 \034\351i\034\337?\031\330\b\210\320\033\211\036\02008\362E\202\341fI\350b|5\231\225\202\004\317\267@8\361\203\247\330hR\\\v\003/^\245\032\220\210\036\217\267;\2446\207\365\022\033Z\203\302P\375K\363A\367\377\356t\334B\0178\375\342\262kw|\333\355+>\315\262\350\320\341\226>{ \316\256\237\311\227\265\027\"1A\3111\373\343\310\377\276?h)\200+\226,\357%\262\333HK\317\314:[\221,.\343FJ\345\210\322j\364\362~\252\v\344\r\374#\330:H\334)\021\254\377\232\026\vx\300\025~Nb\362\241\241\362\244\204\343\205$\311&\"\203\232:\361\333J4\032p\274\177>", 0x185)		 = 389 0
 19246/100087:  gettimeofday(0x7FFFFFFFAAD0, 0x0, 0x0)		 = 0 0
 19246/100087:  clock_gettime(0x4, 0x7FFFFFFFAAD0, 0x0)		 = 0 0
 19246/100087:  getpid(0x0, 0x0, 0x0)		 = 19246 0
 19246/100087:  write(0x7, "\025\003\003\0", 0x45)		 = 69 0
 19246/100087:  shutdown(0x7, 0x1, 0x0)		 = 0 0
 19246/100087:  close(0x7)		 = 0 0
 19246/100087:  clock_gettime(0x4, 0x7FFFFFFFAC70, 0x0)		 = 0 0
 19246/100087:  kevent(0x9, 0x803420A80, 0x2)		 = 2 0
 19246/100087:  clock_gettime(0x4, 0x7FFFFFFFAC70, 0x0)		 = 0 0
```

**Comment by bsuh** [comment]
It seems like socket is not fully connected when nghttpx calls getsockopt and then writev fails and nghttpx calls close. It may be a bug in libev.

**Comment by bsuh** [comment]
```
Breakpoint 2, shrpx::HttpDownstreamConnection::attach_downstream (
    this=0x803482980, downstream=0x80348c900)
    at shrpx_http_downstream_connection.cc:204
204	      rv = connect(conn_.fd, &addr.addr.su.sa, addr.addr.len);
Current language:  auto; currently c++
(gdb) c
Continuing.

Breakpoint 7, ev_feed_event (loop=0x8013ca7f8, w=0x803482a40, revents=2)
    at ev.c:1988
1988	  W w_ = (W)w;
Current language:  auto; currently minimal
(gdb) bt
#0  ev_feed_event (loop=0x8013ca7f8, w=0x803482a40, revents=2) at ev.c:1988
#1  0x000000000055dcfa in shrpx::HttpDownstreamConnection::signal_write (
    this=0x803482980) at shrpx_http_downstream_connection.cc:1094
#2  0x000000000055dc85 in shrpx::HttpDownstreamConnection::push_request_headers
    (this=0x803482980) at shrpx_http_downstream_connection.cc:459
#3  0x00000000004c33b3 in shrpx::Downstream::push_request_headers (
    this=0x80348c900) at shrpx_downstream.cc:520
#4  0x00000000004bcac5 in shrpx::(anonymous namespace)::htp_hdrs_completecb (
    htp=0x8034177a0) at shrpx_https_upstream.cc:406
#5  0x000000000057885f in http_parser_execute (parser=0x8034177a0,
    settings=0x7c5230,
```

Write is signaled from HttpDownstreamConnection::push_request_headers before socket is connected triggering connectcb.

**Comment by tatsuhiro-t** [comment]
Awesome.  Although I still could not reproduce this issue without any logging, I think this is cause of the problem because it explains the issue well.

**Comment by tatsuhiro-t** [comment]
@bstst v1.10.0 contains the fix.  Could you test it?

**Comment by bstst** [comment]
@tatsuhiro-t, yes, looks like the fix does indeed fixes the issue. No more random 502s. Thank you!
